apps/omg_performance/lib/omg_performance/byzantine_events.ex

```diff
@@ -0,0 +1,146 @@
+  """
+  @spec start_dos_get_exits([non_neg_integer()], [OMG.TestHelper.entity()], watcher_url: binary()) ::
+          stats_t()
+  def start_dos_get_exits(positions, dos_users, url \\ @watcher_url) do
+    1..dos_users
+    |> Enum.map(fn _ -> DoSExitWorker.get_exits_fun(positions, url) |> Task.async() end)
+    |> Enum.map(&compute_std_exits_statistics/1)
+  end
+
+  @doc """
+  Fetches utxo positions for a given users list.
+  """
+  @spec get_exitable_utxos([%{addr: binary()}], watcher_url: binary()) :: [non_neg_integer()]
+  def get_exitable_utxos(entities, watcher_url \\ @watcher_url)
+
+  def get_exitable_utxos(addr, watcher_url) when is_binary(addr) do
+    {:ok, utxos} = WatcherClient.get_exitable_utxos(addr, watcher_url)
+    utxos
+  end
+
+  def get_exitable_utxos(%{addr: addr}, watcher_url) when is_binary(addr),
+    do: Encoding.to_hex(addr) |> get_exitable_utxos(watcher_url)
+
+  def get_exitable_utxos(users, watcher_url) when is_list(users),
+    do: Enum.map(users, &get_exitable_utxos(&1, watcher_url)) |> Enum.concat()
+
+  def watcher_synchronize(watcher_url \\ @watcher_url) do
+    Eth.WaitFor.repeat_until_ok(fn -> watcher_synchronized?(watcher_url) end)
+  end
+
+  def watcher_synchronize_service(expected_service, min_service_height, watcher_url \\ @watcher_url) do
+    Eth.WaitFor.repeat_until_ok(fn ->
+      with {:ok, %{services_synced_heights: services_synced_heights}} <- WatcherClient.get_status(watcher_url),
+           %{"height" => height} when height >= min_service_height <-
+             Enum.find(services_synced_heights, &match?(%{"service" => ^expected_service}, &1)) do
+        {:ok, height}
+      else
+        _ -> :repeat
+      end
+    end)
+  end
+
+  defp valid_exit_data({:ok, response}), do: valid_exit_data(response)
+  defp valid_exit_data(%{proof: _}), do: true
+  defp valid_exit_data(_), do: false
+
+  defp compute_std_exits_statistics(task) do
+    {time, exits} = Task.await(task, :infinity)
+    valid? = Enum.map(exits, &valid_exit_data/1)
+
+    %{
+      span_ms: div(time, @micros_in_millisecond),
+      corrects_count: Enum.count(valid?, & &1),
+      errors_count: Enum.count(valid?, &(!&1))
+    }
+  end
+
+  defp watcher_synchronized?(watcher_url) do
+    # Tricky part that deserves a note. This function is prepared to be called in `WaitFor.repeat_until_ok`.
+    # It repeatedly ask for Watcher's `/status.get` until:
+    #  1. last_mined_child_block_number == last_validated_child_block_number, so Watcher synced to last ch-ch block
+    #  2. last_validated_child_block_number > 0, and we expect there will be at least one block as it's called
+    # after perftest.
+    with {:ok,
+          %{
+            last_mined_child_block_number: last_validated_child_block_number,
+            last_validated_child_block_number: last_validated_child_block_number
+          }}
+         when last_validated_child_block_number > 0 <- WatcherClient.get_status(watcher_url) do
```

Comment:
is there a reason why you're avoiding case statements? it is difficult to notice that you're in fact expecting `last_mined_child_block_number == last_validated_child_block_number`.

correct this so that you use a pinned variable with a case statement if you want to use pattern matching.